lib/Service/OnedriveStorageAPIService.php

```diff
@@ -294,75 +316,71 @@ private function downloadDir(string $userId, Folder $topFolder,
-					}
-					if (!is_null($maxDownloadSize) && $newDownloadedSize >= $maxDownloadSize) {
-						throw new MaxDownloadSizeReachedException('Yep');
+					if ($size !== null) {
+						$newDownloadedSize += $size;
+						if ($size > 0) {
+							$newNbDownloaded++;
+							$this->config->setUserValue($userId, Application::APP_ID, 'imported_size', (string)($alreadyImportedSize + $newDownloadedSize));
+							$this->config->setUserValue($userId, Application::APP_ID, 'nb_imported_files', (string)($alreadyImportedNumber + $newNbDownloaded));
+							$this->config->setUserValue($userId, Application::APP_ID, 'last_onedrive_import_timestamp', (string)(new \DateTime())->getTimestamp());
+						}
+						if ($maxDownloadSize !== null && $newDownloadedSize >= $maxDownloadSize) {
+							throw new MaxDownloadSizeReachedException('Download size limit reached');
+						}
 					}
 				}
-			}
-			// REMOVE this directory from unfinished list
-			if (array_key_exists($path, $importTree)) {
-				unset($importTree[$path]);
-			}
-			if (isset($result['@odata.nextLink'])
-				&& $result['@odata.nextLink']
-				&& preg_match('/\$skiptoken=/i', $result['@odata.nextLink'])) {
-				$params['$skiptoken'] = preg_replace('/.*\$skiptoken=/', '', $result['@odata.nextLink']);
-			}
-		} while (isset($result['@odata.nextLink']) && $result['@odata.nextLink']);
-
-		// DIRECTORIES
-		$params = [
-			'filter' => 'folder ne null',
-		];
-		do {
-			$result = $this->onedriveApiService->request($userId, $endPoint, $params);
-			if (isset($result['error']) || !isset($result['value']) || !is_array($result['value'])) {
-				return [
-					'downloadedSize' => $newDownloadedSize,
-					'totalSeenNumber' => $newTotalSeenNumber,
-					'nbDownloaded' => $newNbDownloaded,
-				];
-			}
-			// store directories we have to do
-			foreach ($result['value'] as $item) {
+				// folders: remember for recursion
 				if (isset($item['folder'])) {
-					$subDirPath = $path . '/' . $item['name'];
-					$importTree[$subDirPath] = 'todo';
+					$subDirs[] = $item['name'];
+					// mark for progress tracking
+					$subPath = ltrim($path . '/' . $item['name']);
+					$importTree[$subPath] = 'todo';
 				}
 			}
-			// then explore sub directories
-			foreach ($result['value'] as $item) {
-				if (isset($item['folder'])) {
-					$subDownloadResult = $this->downloadDir(
-						$userId, $topFolder, $maxDownloadSize, (int)$newDownloadedSize, (int)$newTotalSeenNumber, (int)$newNbDownloaded,
-						$path . '/' . $item['name'], $alreadyImportedSize, $alreadyImportedNumber, $importTree
-					);
-					$newDownloadedSize = $subDownloadResult['downloadedSize'];
-					$newTotalSeenNumber = $subDownloadResult['totalSeenNumber'];
-					$newNbDownloaded = $subDownloadResult['nbDownloaded'];
-				}
+
+			// if this directory was marked unfinished, remove it now
+			if (isset($importTree[$path])) {
+				unset($importTree[$path]);
```

Comment:
Is this the classic case or an edge case?
I mean, is it really possible that the current dir is not marked as unfinished?